File: Mathlab/Acceleration_velocity_position_1234/Exercise_1.m
clear
close all

%GIVEN
T1 = 1
T2 = 2
T3 = 3
amax = 2
T = 4 * T1 + 2 * T2 + T3

t1 = T1
t2 = T1 + T2
t3 = T1 + T2 + T1
t4 = T1 + T2 + T1 + T3
t5 = T1 + T2 + T1 + T3 + T1
t6 = T1 + T2 + T1 + T3 + T1 + T2
t7 = T1 + T2 + T1 + T3 + T1 + T2 + T1


dt = T / 1000
t = 0:dt:T;


a1 = amax * t
a2 = amax
a3 = -amax * t + 8
a4 = 0
a5 = -amax * t + 14
a6 = -amax
a7 = amax * t - 22

N=length(t)
vk=zeros(1,N);
sk=zeros(1,N);
vk(1)=0;
sk(1)=0;

a = a1.*(t<=t1) + a2.*((t>t1)&(t<t2)) + a3.*((t>t2)&(t<t3)) + a4.*((t>t3)&(t<t4)) + a5.*((t>t4)&(t<t5)) + a6.*((t>t5)&(t<t6)) + a7.*((t>t6)&(t<t7));

for n=1:N-1
    vk(n+1)=vk(n)+a(n)*dt;
    sk(n+1)=sk(n)+vk(n)*dt+1/2*a(n)*dt^2;
end

figure(4)
subplot(3,1,1)
plot(t,a,'linewidth',1.5)
xlim([0, T])
grid
title('kiihtyvyys a(t)')
subplot(3,1,2)
plot(t,vk,'linewidth',1.5)
xlim([0, T])
title('nopeus v(t)')
grid
subplot(3,1,3)
plot(t,sk,'linewidth',1.5)
xlim([0, T])
grid
title('paikka s(t)')
xlabel('aika t')

File: AutomationTechnologyMath/DFT/Exercise1.py
import numpy as np
import matplotlib.pyplot as plt

#Given

A_ft = 3
# First case
#f_ft = 22
# Second case
f_ft = 20
phi_ft = 0
N_ft = 20

fs= 100 #sampling frequency

N = 20
#Solve
t=np.linspace(0,T,1000)

#First case
# y = np.array([3, 0.5, -2.8, -1.5, 2.2, 2.4, -1.2, -2.9, 0.2, 3, 0.9, -2.7, -1.9, 1.9, 2.6, -0.9, -3, -0.2, 2.9, 1.3])
# Second case
y = np.array([3, 0.9, -2.4, -2.4, 0.9, 3, 0.9, -2.4, -2.4, 0.9, 3, 0.9, -2.4, -2.4, 0.9, 3, 0.9, -2.4, -2.4, 0.9])
M = int(N / 2)
n = np.arange(0, N)

a0 = 1 / N * np.sum(y)
a = np.zeros(M)
b = np.zeros(M)

from numpy import cos, sin, pi
for k in range(1,M):
    a[k-1]=2/N*np.sum(y*cos(2*pi*k*n/N))
    b[k-1]=2/N*np.sum(y*sin(2*pi*k*n/N))

a[M-1]=1/N*np.sum(y*cos(2*pi*M*n/N))
b[M-1]=1/N*np.sum(y*sin(2*pi*M*n/N))

A=np.sqrt(a**2+b**2) #amplitudes
phi=np.arctan2(-b,a) #phase angles

f0=fs/N #basic frequency
f=np.arange(f0,(M+1)*f0,f0) #frequencies

T=1/f0
ts=n/fs
print(ts)

t=np.linspace(0,T,1000)

plt.figure(figsize=(7,4))

yt=a0
for k in range(M):
    yt=yt+A[k]*cos(2*pi*f[k]*t+phi[k])

plt.subplot(211)
plt.plot(t, A_ft * cos(2 * pi * f_ft * t + phi_ft),lw=2,label='f(t)')
plt.plot(t, yt,'orange', label = 'y(t)',lw=2)
plt.plot(ts,y,'ro')
plt.grid()
plt.legend(fontsize = 14)
plt.xlabel('time',fontsize=14)
plt.title('A = ' + str(A_ft) + ', f = '+ str(f_ft) + ', $\phi$ = '+str(phi_ft) + ', N = ' + str(N_ft)+ ', $f_s$ = '+ str(fs))

plt.subplot(212)
plt.bar(np.hstack((0,f)),np.hstack((a0,A)),edgecolor='k',zorder=2)
plt.grid()
plt.xticks(np.hstack((0,f)))
plt.xlabel("frequency")
plt.ylabel("amplitudes")
plt.show()
